### ISSUE TITLE
Release candidate JSXGraph v1.5.0rc1

### ISSUE THREAD

**Issue by alfredwassermann** [jsxgraph/jsxgraph#494]
Testing the release candidate v1.5.0rc1 would be highly appreciated, see [the blog entry on the JSXGraph page](https://jsxgraph.org/wp/2022-12-19-release-candidate-1.5.0/) for more information. Please, download the files from

- <https://jsxgraph.org/distrib/jsxgraphcore_1.5.0rc1.js> minified version
- <https://jsxgraph.org/distrib/jsxgraph_1.5.0rc1.css> CSS file
- <https://jsxgraph.org/distrib/jsxgraphsrc_1.5.0rc1.js> unminified version for development

and give us feedback below.

**Comment by geometryzen** [comment]
Alfred,

Congratulations on moving the jsxgraph source code towards ES modules.

However, I am encountering some difficulties using the 1.5.0 rc1.

The following import statement does not work...

import { JSXGraph } from 'jsxgraph';

Build tools like webpack and rollup will tell you that the export 'JSXGraph' was not found in 'jsxgraph'.

I did find that the following default import does work...

import JXG from 'jsxgraph'.

I also had success using the UMD format jsxgraphcore.js bundle.

It appears to me that the jsxgraph source code is not structured to produce individual exports and is only exporting a single JXG object.

Am I understanding this correctly?

David

P.S.

The full backstory here is that I was trying to create a bundled version of jsxgraph to create both ES and System formats using rollup. This is something that I have done successfully before with other packages. As a side note, the canvas dependency, which is not required for the browser, creates some issues for rollup.

**Comment by alfredwassermann** [comment]
There is second release candidate available to test: Please, download the files from

    https://jsxgraph.org/distrib/jsxgraphcore_1.5.0rc2.js minified version
    https://jsxgraph.org/distrib/jsxgraph_1.5.0rc2.css CSS file
    https://jsxgraph.org/distrib/jsxgraphsrc_1.5.0rc2.js unminified version for development

and give us feedback below.

**Comment by alfredwassermann** [comment]
@geometryzen thank you very much for your detailed analysis. You stated two problems: 

1) *`import { JSXGraph } from 'jsxgraph';` does not work.*
I'm not sure if we should support that. At the moment, the way to use JSXGraph in a module is:

```
import JXG from 'jsxgraph';
const board = JXG.JSXGraph.initBoard(...);
```
We are open to support also `import { JSXGraph } from 'jsxgraph';` , but what should be the content of `JSXGraph` ?

2) *It appears to me that the jsxgraph source code is not structured to produce individual exports and is only exporting a single JXG object.*

First, the annoying canvas dependency should now be resolved, see 1.5.0-rc2. 
Second, from 1.5.0+ jsxgraphcore.js supports the classical loading in the web browser with `<script>`, AMD, import as a CommonJS module. Import as an ES6 module is enabled with `jsxgraphcore.mjs`. Both, `jsxgraphcore.js` and `jsxgraphcore.mjs`  are created with webpack. The control file for the bundle is `src/index.js`. Isn't it possible to use a similar approach with rollup?

Best wishes,
Alfred

**Comment by geometryzen** [comment]
@alfredwassermann

Sorry I can't be more directly involved with PRs at the moment, but I hope I can still be of service.

1) The granularity issue is a breaking change for me, although my perspective might be a bit of an edge case. With jsxgraph@1.4.x it was possible to load the jsxgraphcore.js UMD file with a SystemJS loader and then use the granular syntax form:

import { JSXGraph } from 'jsxgraph'

I've been doing this for a long time in STEMCstudio, and try to avoid the JSX prefix approach which seems more appropriate when jsxgraph is loaded as a global variable.

Ultimately, I think this is what you would want to offer as it is better from a tree-shaking perspective.

An approach to getting back the granularity without internal refactoring is to do something like the following in your top level index.js file.

// Values
export const JSXGraph = JXG.JSXGraph;
// etc...

// Classes
export const Board = JXG.Board;
// etc...

// Functions
export const LMS2rgb = JXG.LMS2rgb;
// etc

Over time you could refactor internally to make the granular objects the primary objects and then assemble them onto JXG as a secondary step. In fact, this second step would usually take place in the build.

For the time being I'm taking this approach in https://github.com/geometryzen/esm-bundle-jsxgraph. This repository exists because I want a bundled system format, but I'm also using it to workaround the granularity issue. Ideally I'd like to simply have:

export {*} from 'jsxgraph'

I can help you build this complete list to achieve granularity.

2) It would be good to add the more modern "exports" property to your package.json.

"exports": {
   ".": {
       "types": "./src/index.d.ts",
       "default: "./src/index.js"
}
}

3) I'm a bit "concerned" about the way that 1.5.x module code is modifying the global JXG variable as a side-effect. It's not a deal-breaker, but it doesn't seem optimal. A better approach would be to make this happen only if you use the UMD format of jsxgraph in a global (non-AMD, non-require) setting.

Best Wishes,

David

P.S. Would be glad to zoom to discuss any of this. I also have a request to generalize initBoard to take either and elementId or an HTMLDivElement as a way to support JsxGraph in a Web Component shadow DOM.

**Comment by alfredwassermann** [comment]
@geometryzen Thank you very much for your help. It is very important for JSXGraph to make the transition to ES6 as backward compatible as possible. Where are we now?

- It seems that at the time being, the webpack build of the file `jsxgraphcore.js` is fully equivalent to the 1.4 branch, additionally it got rid of the canvas (and any other) dependency.
- When it comes to bundling the ES6 source code of JSXGraph you pointed out the incompatibilities to the UMD version. I'm not sure if I understand your suggestions about future refactoring steps. Maybe my last commits go in this direction: I added a file `index2.js` and refactored three files: `base/element.js`, `base/curve.js` and `base/constants.js`. The bundling with webpack using this file `index2.js` gives the same result. Additionally,  one can import individual objects like

```
import { GeometryElement, Curve, COORDS_BY_SCREEN } from '.... src/index2.js';
console.log(GeometryElement, Curve, COORDS_BY_SCREEN);
```

The advantage would be that the exports of each object are handled locally in its file like `curve.js` and not in `index.js`. What do you think? Is it what you meant?

I do not understand the "side-effects" you mention in 3. 

Zoom would we great! Are you available, let's say on Wednesday? For this, I suggest to switch to email.

Best wishes,
Alfred

**Comment by stemcstudio** [comment]
Alfred,

I haven't really tried using the jsxgraphcore.js @ 1.5.x, but I can do that and get back to you.

What you are doing in index2.js corresponds with what I was expecting. Of course, you can do it in index.js incrementally with the following changes:

// export const Curve = JXG.Curve;
export { Curve } from "./base/curve";

In my humble opinion, using the explicit export over "export * from" feels a bit safer because I can really see what is going on at the API level.

Looking now at curve.js, in my experience, using default exports is not worth the apparent saving in curly braces. Additionally, exporting Curve as both the default export and a named export is probably not worth the duplication. So I would simplify

const Curve = JXG.Curve;
export { Curve as default, Curve};

to

export { Curve };

And assuming that you also migrate you code to classes (over time), this will become...

export class Curve {
}
JXG.Curve = Curve;

Eventually, you will be able to get rid of that last assignment; you just won't need it. I see now that JXG is imported from "./jxg" so it is only a module level variable at this stage.

The "side-effect" I was referring to is in index.js:

if (Env.isBrowser) {
    window.JXG = JXG;
} else if (Env.isWebWorker()) {
    self.JXG = JXG;
}

By virtue of consuming the index.js module, the consumer is having additions made to either window or self. Whether this is a good idea or not is a bit subtle because... the index.d.ts says you can work with both modules and a namespace variable. But that had to be the case to support the evolution from global namespaces to modules. Two wrongs won't make a right! The JXG namespace in index.d.ts should be @deprecated and the window.JXG = JXG assignment can be removed when the deprecation is made followed through.

The bottom line is that I would not worry about the JXG global for 1.5.0 and your current path is safe for users. Technically, removing JXG could be regarded as a backwardly incompatible API change best reserved for jsxgraph@2.0? Maybe!

I hope that wasn't too confusing.

I could zoom tomorrow, but it would be preferable (for me) to see how you feel about this feedback and leave talking as an option for next week if you still think it would help.

BTW. If you would do a 1.5.0-rc3 with the export changes already done then I can verify that.

Best Wishes,

David

**Comment by stemcstudio** [comment]
Alfred,

I just tried using jsxgraphcore.js from the 1.5.0-rc2 branch. I haven't (and can't) do a thorough test of all the code, but the basic functionality seems to work fine.

David

**Comment by stemcstudio** [comment]
Alfred,

With a bit of experimentation, I found that the following change to index.d.ts will deprecate the JXG namespace without affecting use of the granular objects.

/**
 * @deprecated
 */
 declare namespace JXG {
    export interface Graph {
        initBoard(): void
        freeBoard(): void
    }
    const JSXGraph: Graph;
}

declare module "jsxgraph" {
    export = JXG    
}

David
